test/persistence-hooks.suite.js

```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+              observedContexts.should.eql(aTestModelCtx({
+                data: {
+                  id: 'new-id',
+                  name: 'a name'
+                },
+                isNewInstance: true,
+                currentInstance: {
+                  id: 'new-id',
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
+            }
+            done();
+          });
+      });
+
+      it('triggers `persist` hook on replace', function(done) {
+        TestModel.observe('persist', pushContextAndNext());
+
+        TestModel.replaceOrCreate(
+          { id: existingInstance.id, name: 'replaced name' },
+          function(err, instance) {
+            if (err) return done(err);
+            
+            var expected = {
+              where: { id: existingInstance.id },
+              data: {
+                id: existingInstance.id,
+                name: 'replaced name'
+              },
+              currentInstance: {
+                id: existingInstance.id,
+                name: 'replaced name',
+                extra: undefined
+              }
+            };
+
+            var expectedContext;
+            if (dataSource.connector.replaceOrCreate) {
+              expectedContext = aTestModelCtx(expected);
+            } else {
+              expected.isNewInstance = false;
+              expectedContext = aTestModelCtx(expected);       
+          }
+
+            observedContexts.should.eql(expectedContext);
+            done();
+          });
+      });
+
+      it('triggers `loaded` hook on create', function(done) {
+        TestModel.observe('loaded', pushContextAndNext());
+
+        TestModel.replaceOrCreate(
+          { id: 'new-id', name: 'a name' },
+          function(err, instance) {
+            if (err) return done(err);
+            observedContexts.should.eql(aTestModelCtx({
+              data: {
+                id: 'new-id',
+                name: 'a name'
+              },
+              isNewInstance: true
+            }));     
+            done();
+          });
+      });
+
+      it('triggers `loaded` hook on replace', function(done) {
+        TestModel.observe('loaded', pushContextAndNext());
+
+        TestModel.replaceOrCreate(
+          { id: existingInstance.id, name: 'replaced name' },
+          function(err, instance) {
+            if (err) return done(err);
+
+            if (dataSource.connector.replaceOrCreate) {
+              observedContexts.should.eql(aTestModelCtx({
+                data: {
+                  id: existingInstance.id,
+                  name: 'replaced name'
+                },
+                isNewInstance: false
+              }));
+            } else {
+              // For Unoptimized (non-atomic) connector, the callback
+              // function `pushContextAndNext` is called twice.
+              // As a result, observedContexts returns an array and
+              // NOT a single instance.
```

Comment:
As commented before, I would like to understand why the "loaded" hook is triggered twice and check if there is a reasonably easy solution that would allow us to have it triggered only once.